File: modern-cms-platform-pro/src/controllers/ContentController.cc
#include "ContentController.h"
#include "services/ContentService.h"
#include "utils/ApiResponse.h"
#include "models/User.h" // To get user_id from req context

void ContentController::getAllContent(const drogon::HttpRequestPtr &req,
                                       std::function<void(const drogon::HttpResponsePtr &)> &&callback) {
    LOG_INFO << "Fetching all content.";
    ContentService::getInstance()->getAllContent(
        [callback](const std::vector<Content>& contents, const std::string& error) {
            if (!error.empty()) {
                LOG_ERROR << "Failed to retrieve content: " << error;
                return callback(ApiResponse::internalError("Failed to retrieve content."));
            }
            Json::Value data = Json::arrayValue;
            for (const auto& c : contents) {
                data.append(c.toJson());
            }
            return callback(ApiResponse::ok("Content retrieved successfully.", data));
        }
    );
}

void ContentController::getContentById(const drogon::HttpRequestPtr &req,
                                        std::function<void(const drogon::HttpResponsePtr &)> &&callback,
                                        const std::string &id) {
    LOG_INFO << "Fetching content by ID: " << id;
    ContentService::getInstance()->getContentById(id,
        [callback](const std::optional<Content>& content, const std::string& error) {
            if (!error.empty()) {
                LOG_ERROR << "Failed to retrieve content by ID: " << error;
                return callback(ApiResponse::internalError("Failed to retrieve content."));
            }
            if (!content) {
                return callback(ApiResponse::notFound("Content not found."));
            }
            return callback(ApiResponse::ok("Content retrieved successfully.", content->toJson()));
        }
    );
}

void ContentController::createContent(const drogon::HttpRequestPtr &req,
                                       std::function<void(const drogon::HttpResponsePtr &)> &&callback) {
    LOG_INFO << "Creating new content.";
    auto json = req->getJsonObject();
    if (!json || !json->isMember("title") || !json->isMember("body")) {
        return callback(ApiResponse::badRequest("Missing title or body."));
    }

    // Get author_id from request context (set by AuthMiddleware)
    auto user = req->getAttributes()->get<User>("user_info");
    if (user.id.empty()) {
        return callback(ApiResponse::unauthorized("User information not found in token."));
    }

    Content content;
    content.title = (*json)["title"].asString();
    content.body = (*json)["body"].asString();
    content.author_id = user.id;
    content.slug = drogon::utils::urlEncode(content.title); // Simple slug generation
    if (json->isMember("status")) {
        content.status = (*json)["status"].asString();
    } else {
        content.status = "draft";
    }

    ContentService::getInstance()->createContent(content,
        [callback](const std::optional<Content>& newContent, const std::string& error) {
            if (!error.empty()) {
                LOG_ERROR << "Failed to create content: " << error;
                return callback(ApiResponse::internalError("Failed to create content."));
            }
            if (!newContent) {
                 return callback(ApiResponse::internalError("Failed to create content, no content returned."));
            }
            return callback(ApiResponse::created("Content created successfully.", newContent->toJson()));
        }
    );
}

void ContentController::updateContent(const drogon::HttpRequestPtr &req,
                                       std::function<void(const drogon::HttpResponsePtr &)> &&callback,
                                       const std::string &id) {
    LOG_INFO << "Updating content with ID: " << id;
    auto json = req->getJsonObject();
    if (!json) {
        return callback(ApiResponse::badRequest("Invalid JSON payload."));
    }

    // Get author_id from request context for authorization check
    auto user = req->getAttributes()->get<User>("user_info");
    if (user.id.empty()) {
        return callback(ApiResponse::unauthorized("User information not found in token."));
    }

    ContentService::getInstance()->updateContent(id, user.id, *json,
        [callback](const std::optional<Content>& updatedContent, const std::string& error) {
            if (!error.empty()) {
                if (error == "Not Found") {
                    return callback(ApiResponse::notFound("Content not found or not authorized to update."));
                }
                LOG_ERROR << "Failed to update content: " << error;
                return callback(ApiResponse::internalError("Failed to update content."));
            }
            if (!updatedContent) {
                 return callback(ApiResponse::internalError("Failed to update content, no content returned."));
            }
            return callback(ApiResponse::ok("Content updated successfully.", updatedContent->toJson()));
        }
    );
}

void ContentController::deleteContent(const drogon::HttpRequestPtr &req,
                                       std::function<void(const drogon::HttpResponsePtr &)> &&callback,
                                       const std::string &id) {
    LOG_INFO << "Deleting content with ID: " << id;
    // Get author_id from request context for authorization check
    auto user = req->getAttributes()->get<User>("user_info");
    if (user.id.empty()) {
        return callback(ApiResponse::unauthorized("User information not found in token."));
    }

    ContentService::getInstance()->deleteContent(id, user.id,
        [callback](bool success, const std::string& error) {
            if (!error.empty()) {
                if (error == "Not Found") {
                    return callback(ApiResponse::notFound("Content not found or not authorized to delete."));
                }
                LOG_ERROR << "Failed to delete content: " << error;
                return callback(ApiResponse::internalError("Failed to delete content."));
            }
            if (!success) {
                 return callback(ApiResponse::internalError("Failed to delete content."));
            }
            return callback(ApiResponse::noContent());
        }
    );
}

File: production-cms-backend-complete/backend/src/controllers/ContentController.ts
import { Request, Response, NextFunction } from 'express';
import { ContentService } from '../services/ContentService';
import { CustomError } from '../middlewares/errorHandler';
import { CreateContentDto, UpdateContentDto } from '../utils/validation';
import { ContentStatus } from '../entities/Content';

export class ContentController {
    private contentService = new ContentService();

    async createContent(req: Request, res: Response, next: NextFunction) {
        try {
            if (!req.user) {
                return next(new CustomError(401, "User not authenticated for content creation."));
            }
            const contentData: CreateContentDto = req.body;
            const newContent = await this.contentService.createContent(contentData, req.user.id);
            res.status(201).json({
                success: true,
                message: 'Content created successfully',
                data: newContent,
            });
        } catch (error) {
            next(error);
        }
    }

    async getContents(req: Request, res: Response, next: NextFunction) {
        try {
            const page = parseInt(req.query.page as string) || 1;
            const limit = parseInt(req.query.limit as string) || 10;
            const status = req.query.status as ContentStatus | undefined;
            const search = req.query.search as string | undefined;

            const [contents, total] = await this.contentService.getContents(page, limit, status, search);
            res.status(200).json({
                success: true,
                data: contents,
                meta: {
                    page,
                    limit,
                    total,
                    totalPages: Math.ceil(total / limit),
                },
            });
        } catch (error) {
            next(error);
        }
    }

    async getContentById(req: Request, res: Response, next: NextFunction) {
        try {
            const { id } = req.params;
            const content = await this.contentService.getContentById(id);
            if (!content) {
                return next(new CustomError(404, 'Content not found.'));
            }
            res.status(200).json({
                success: true,
                data: content,
            });
        } catch (error) {
            next(error);
        }
    }

    async updateContent(req: Request, res: Response, next: NextFunction) {
        try {
            if (!req.user) {
                return next(new CustomError(401, "User not authenticated for content update."));
            }
            const { id } = req.params;
            const updateData: UpdateContentDto = req.body;
            const updatedContent = await this.contentService.updateContent(id, updateData, req.user.id, req.user.role);
            if (!updatedContent) {
                return next(new CustomError(404, 'Content not found.'));
            }
            res.status(200).json({
                success: true,
                message: 'Content updated successfully',
                data: updatedContent,
            });
        } catch (error) {
            next(error);
        }
    }

    async deleteContent(req: Request, res: Response, next: NextFunction) {
        try {
            if (!req.user) {
                return next(new CustomError(401, "User not authenticated for content deletion."));
            }
            const { id } = req.params;
            await this.contentService.deleteContent(id, req.user.id, req.user.role);
            res.status(204).json({
                success: true,
                message: 'Content deleted successfully',
            });
        } catch (error) {
            next(error);
        }
    }
}